src/empirical.jl

```diff
@@ -51,18 +51,43 @@ struct EmpiricalVariogram
     sums   = zeros(nlags)
     counts = zeros(Int, nlags)
 
+    # decide whether to construct a spatial index
+    # coordinates need to be AbstractFloat for NearestNeighbors to work
+    # https://github.com/KristofferC/NearestNeighbors.jl/issues/13
+    isfloat = eltype(coordinates(sdata)) <: AbstractFloat
+    # quite arbitrary, but for small datasets setting up the index is too much overhead
+    # it really also depends on what fraction of points are within hmax
+    isbig = npts > 1000
+
+    if isfloat && isbig
+      neigh = BallNeighborhood(hmax, distance)
+      neighsearcher = NeighborhoodSearcher(sdata, neigh)
+      use_spatial_index = true
+    else
+      use_spatial_index = false
+    end
+
     # preallocate memory for coordinates
     xi = MVector{ndims(sdata),coordtype(sdata)}(undef)
     xj = MVector{ndims(sdata),coordtype(sdata)}(undef)
 
-    @inbounds (
-    for j in 1:npts
+    @inbounds for j in 1:npts
       coordinates!(xj, sdata, j)
-      for i in j+1:npts
+
+      if use_spatial_index
+        pt_indices = search(xj, neighsearcher)
+      else
+        pt_indices = j+1:npts
+      end
+
+      for i in pt_indices
+        i <= j && continue  # spatial index include these
         coordinates!(xi, sdata, i)
 
         # evaluate lag and (cross-)variance
         h = evaluate(distance, xi, xj)
+        # early exit if out of range
+        h > hmax && continue
```

Comment:
That is a nice catch. We should benchmark it to see if the check is faster than computing the floating point `v` regardless.

Reply:
Interestingly the `v` calculation is the most expensive, it spends about 50% of its time there. So avoiding it pays off, especially if it is true often. Forcing it manually to take the no spatial index branch, this:

```julia
using BenchmarkTools
using GeoStats

n = 10_000
X = rand(2, n)
z = rand(n)

@btime EmpiricalVariogram($X, $z, nlags=20, maxlag=0.05)
```

Will show:
```
no spatial index with hmax early exit
1.002 s (4840077 allocations: 74.02 MiB)
no spatial index without hmax early exit
10.080 s (540516782 allocations: 8.05 GiB)
```

Reply:
Nice to see these speedups. 💯